File: examples/sample_java/src/main/java/io/karte/sample_java/MyFirebaseMessagingService.java
package io.karte.sample_java;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import java.util.Map;

import io.karte.android.notifications.MessageHandler;
import io.karte.android.notifications.Notifications;


public class MyFirebaseMessagingService extends FirebaseMessagingService {

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {

        boolean handled = MessageHandler.handleMessage(this, remoteMessage);
        if (!handled) {
            Map<String, String> data = remoteMessage.getData();
            String title = data.get("subject");
            String body = data.get("text");
            sendNotification(title, body, data.get("android_channel_id"), data);
        }
    }

    protected void sendNotification(String title, String body, String channel, Map<String, String> data) {

        Intent intent = new Intent(this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        MessageHandler.copyInfoToIntent(data, intent);

        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent,
                PendingIntent.FLAG_ONE_SHOT);

        Uri defaultSoundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);

        NotificationCompat.Builder notificationBuilder =
                new NotificationCompat.Builder(this, "my_channel")
                        .setSmallIcon(getApplicationInfo().icon)
                        .setContentTitle(title)
                        .setContentText(body)
                        .setAutoCancel(true)
                        .setSound(defaultSoundUri)
                        .setChannelId(channel)
                        .setContentIntent(pendingIntent);

        NotificationManager notificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        if (notificationManager == null) {
            return;
        }
        notificationManager.notify(0, notificationBuilder.build());
    }

    @Override
    public void onNewToken(@NonNull String token) {
        super.onNewToken(token);
        Notifications.registerFCMToken(token);
    }
}

File: examples/sample_kotlin/src/main/java/io/karte/sample_kotlin/MyFirebaseMessagingService.kt
package io.karte.sample_kotlin

import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.media.RingtoneManager
import androidx.core.app.NotificationCompat
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import io.karte.android.KarteApp
import io.karte.android.notifications.MessageHandler
import io.karte.android.notifications.registerFCMToken

class MyFirebaseMessagingService : FirebaseMessagingService() {

    override fun onMessageReceived(remoteMessage: RemoteMessage) {

        val handled = MessageHandler.handleMessage(this, remoteMessage)
        if (!handled) {
            val data = remoteMessage.data
            val title = data["subject"]
            val body = data["text"]
            sendNotification(title, body, data.get("android_channel_id"))
        }
    }

    private fun sendNotification(
        title: String?,
        body: String?,
        channel: String?
    ) {

        val intent = Intent(this, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)

        val pendingIntent = PendingIntent.getActivity(
            this, 0, intent,
            PendingIntent.FLAG_ONE_SHOT
        )

        val defaultSoundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)

        val notificationBuilder = NotificationCompat.Builder(this, "my_channel")
            .setSmallIcon(getApplicationInfo().icon)
            .setContentTitle(title)
            .setContentText(body)
            .setAutoCancel(true)
            .setSound(defaultSoundUri)
            .setContentIntent(pendingIntent)
        channel?.let { notificationBuilder.setChannelId(it) }

        val notificationManager =
            getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
                ?: return

        notificationManager.notify(0, notificationBuilder.build())
    }

    override fun onNewToken(token: String) {
        super.onNewToken(token)
        KarteApp.registerFCMToken(token)
    }
}
